Make BeamSearchDecoder compatible with float16

georgesterpu opened the pull request:
Make BeamSearchDecoder compatible with float16 when a non-zero `length_penalty_weight` is used.
Addresses #22172

georgesterpu commented on tensorflow/contrib/seq2seq/python/ops/beam_search_decoder.py:
```diff
@@ -819,6 +819,8 @@ def _get_scores(log_probs, sequence_lengths, length_penalty_weight):
   """
   length_penalty_ = _length_penalty(
       sequence_lengths=sequence_lengths, penalty_factor=length_penalty_weight)
+  if not isinstance(length_penalty_, float):
```
Hi Eugene.
When `length_penalty_` is a Python float, math_ops.cast will cast it into a Tensor. Would this add one more node to the graph, making the default case slightly slower ?
Would it be better to import and use gen_math_ops.cast instead, plus the type checking ?

ebrevdo replied:
it gets converted to a float32 tensor anyway somewhere.  use cast, or use ops.convert_to_tensor(length_penalty_, log_probs.dtype) if you want it to error out if someone provides a float32 *tensor* penalty but does float16 calculations.  that's probably better anyway.

georgesterpu replied:
Thanks for the reply, I'll try it.

tensorflowbutler commented:
Nagging Assignee @yifeif: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @yifeif: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

tensorflowbutler commented:
Nagging Assignee @yifeif: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

dubey committed:
Correctly log default ranking in collective param resolution.

PiperOrigin-RevId: 218907613

Commit:
Introduce a pass to increase the amount of dynamism supported by an XLA cluster

Increases the amount of dynamism representable by XLA clusters by rewriting the
TensorFlow graph.  See the header for a description.

This pass, combined with jit/partially_decluster_pass, reduces the number of
unnecessary cluster recompilations in some common cases.

The CL is organized as follows:

 - cc/framework/scope* and core/graph/node_builder are modified so that new
   nodes can now be automatically put in an XLA cluster using
   Scope::WithXlaCluster.

 - The pass is implemented in jit/increase_dynamism_for_auto_jit_pass.

 - In jit/jit_compilation_pass_registration The new pass is registered to run
   between MarkForCompilationPass and PartiallyDeclusterPass.

PiperOrigin-RevId: 218907734

tensorflower-gardener committed:
Move contrib/nccl to core/nccl.

PiperOrigin-RevId: 218908694

annarev committed:
Internal change

PiperOrigin-RevId: 218909148

alextp committed:
Fixes accidental breakage in tf.function.

PiperOrigin-RevId: 218915897

Commit:
Manually pass `self` to converted functions, which are as of yet come back unbound after conversion.

PiperOrigin-RevId: 218916376

tensorflower-gardener committed:
Enable generated tests, but keep 'sin' disabled.

PiperOrigin-RevId: 218917937

tensorflower-gardener committed:
Merge pull request #23035 from margaretmz:margaret-dcgan

PiperOrigin-RevId: 218917724

aselle committed:
Add github template and request op compatibility bug reports.

New template is for tf lite operation compatibility.
Exporter now prints a message referring to this templat.

PiperOrigin-RevId: 218921105

feihugis committed:
Add a line break when logging the graph definition

k-w-w committed:
Replace calls to export_savedmodel with export_saved_model, and remove strip_default_attr argument.

PiperOrigin-RevId: 218921778

tensorflower-gardener committed:
Fix tensorflow::data::model::SharedStatefield initializer order to avoid warnings.

The fields will be initiwalized in the order in which they are declared in the struct
regardless of the order the initializers are listed, but some compilers warn if the orderings do not match.
This CL makes tham match to avoid the warnings.

PiperOrigin-RevId: 218922131

bixia1 committed:
[XLA:GPU] Remove a map data structure used to support the code generation of
GetTupleElement inside a fused computation.

The map initially records the fusion parameters and their base addresses. Then
during the forward traversal of the fused computation that produces elemental
ir generators, we generate code to calculate the base address for the
GetTupleElement instructions that produce tuple results, and add the
information to the map. As a result, this map records all the tuple producing
instructions whose results can be used in GetTupleElement instructions. During
code generation, if a GetTupleElement instruction produces a non-tuple result,
we look up the map for the base pointer of the tuple to generate code that
produces the instruction result. On the other hand, if a GetTupleELement
instruction produces a tuple result, we do not generate any code as the code
for such instructions is already generated when we produce the map data
structure.

Such a map data structure is not necessary. The base addresses for a parameter
can be derived from the IrArray of the parameter, which can be looked up
through the parameter number. We change the elemental ir emitter for
GetTupleElement instructions to return the base address for the tuple producing
GetTupleElement instructions to avoid looking up the map for such instructions.
In doing that, we also avoid generating code for tuple producing
GetTupleElement instructions duing the phase that produces elemental ir
generators to clean up the code generation for GetTupleElement instructions.

PiperOrigin-RevId: 218922219

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 218922716

aaroey committed:
Refactor TF-TRT C++ API in convert_nodes.cc, and add TrtNodeValidator class
(which will be later used to do compile-time validation for each TF node to
determine whether it's supported by TRT, in cl/217561555).

Most of the changes are not functional changes but are parameter plumbings,
they come from the 3rd item listed below.

Implementation details:

- Added TrtNodeValidator to validate nodes at compile-time to determine whether
  they're TRT compatible. Validation logic is added for ConvertTranspose and
  ConvertReshape, but the real validation will happen in a separate CL.
- Split APIs to manage the storage of TRT_ShapedWeights to a separate helper
  class TrtWeightStore. Both TrtNodeValidator and Converter need that.
- Change signature of the op converters, by putting all arguments into single
  OpConverterParams struct (***this is where most of the changes come from***).
  Also we added TrtWeightStore separately in the struct, and added
  validation_only for used by TrtNodeValidator.

PiperOrigin-RevId: 218923091

er91 committed:
Token output should be added last, after resource variable updates.

PiperOrigin-RevId: 218925739

tensorflower-gardener committed:
Go: Update generated wrapper functions for TensorFlow ops.
PiperOrigin-RevId: 218925946

Commit:
Add CollectiveAllReduceStrategy into the integration test.

PiperOrigin-RevId: 218927502

shivaniag committed:
Merges `register_concrete` with `add_to_graph`.

PiperOrigin-RevId: 218929028

tensorflower-gardener committed:
Add option to run regular NMS in detection post-processing for SSD.

PiperOrigin-RevId: 218929667

ezhulenev committed:
Use const node pointers in ReverseDfs

PiperOrigin-RevId: 218930546

shivaniag committed:
[tf.data] Replaces Defun with defun for remaining dataset related APIs.

PiperOrigin-RevId: 218932490

asimshankar committed:
[Java]: Release 1.12.0-rc2

PiperOrigin-RevId: 218940996

Commit:
[TF:XLA] Bump open source abseil revision to f86f9413856b65afdd61fea938d684b8ab73115a

PiperOrigin-RevId: 218942595

tensorflower-gardener committed:
Add a TODO to support lookups that has different batch sizes.

PiperOrigin-RevId: 218943406

tensorflower-gardener committed:
Internal Change.

PiperOrigin-RevId: 218945568

Commit:
[XLA] Free the previous run's result before running the next iteration in replay_computation.

Otherwise, we will try to allocate the current iteration's result while the
previous iter's result is still live.  If the result is large enough, this can
OOM us.

PiperOrigin-RevId: 218955957

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2018-10-27

PiperOrigin-RevId: 218968305

jianlijianli committed:
Internal change

PiperOrigin-RevId: 219006877

tensorflower-gardener committed:
Add Nearest Neighbor Resize op to TFLite

PiperOrigin-RevId: 219011298

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2018-10-28

PiperOrigin-RevId: 219027205

saeta committed:
[tf.data / Cloud Bigtable]: Improve error messages.

Previously, the error messages generated by the Bigtable client would not be UTF-8, and thus the real errors would be swallowed by UTF8 conversion problems. This change drops the error_details() when converting from gRPC Status objects to TensorFlow status objects, resulting in the true underlying error messages coming through.

PiperOrigin-RevId: 219062740

saeta committed:
[tf.data / Cloud Bigtable]: Avoid underflow.

Previously, because the size_t type is unsigned, if keys_.empty(), we would underflow, instead of returning end_of_sequence, resulting in a runtime crash. This change switches around the arithmetic so that it's robust to underflow.

PiperOrigin-RevId: 219062746

er91 committed:
Remove newly created Switch nodes in FunctionalizeCond.

PiperOrigin-RevId: 219075975

Commit:
Disable IncreaseDynamismForAutoJitPass

It creates slices of index type DT_INT64 which do not have a kernels on GPU.

PiperOrigin-RevId: 219080243

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2018-10-29

PiperOrigin-RevId: 219095478

Commit:
Improve the tests.

PiperOrigin-RevId: 219137954

tensorflower-gardener committed:
Internal build specification change

PiperOrigin-RevId: 219137973

tensorflower-gardener committed:
DepthwiseConv, create dot-product test placeholders.

PiperOrigin-RevId: 219139518

tensorflower-gardener committed:
Increase DepthwiseConv output depth limit from 2048 to 4832.

PiperOrigin-RevId: 219140225

hawkinsp committed:
[TPU] Move methods to map between TensorFlow task/device names and TPU device coordinates out of DeviceAssignment and into Topology.

The mapping is a property of the Topology (which describes the machine and how to address it), not the DeviceAssignment, which describes a choice of devices for a particular job.

PiperOrigin-RevId: 219141156

Commit:
Add optional task_type and task_index parameters to ClusterResolvers

PiperOrigin-RevId: 219144685

shashishekhar committed:
Fix paths after Makefile movement.

PiperOrigin-RevId: 219161339

tensorflower-gardener committed:
Step 5 of "tower" -> "replica" terminology transition for DistributionStrategy.

PiperOrigin-RevId: 219162518

er91 committed:
Functions to extract outside compilation and construct host side graph.

PiperOrigin-RevId: 219163778

caisq committed:
Add example code snippets to tf.contrib.saved_model.(save|load)_keras_model

PiperOrigin-RevId: 219168444

tensorflower-gardener committed:
Made the XlaBuilder build identical protos when the same esact sequence of instructions have been used to create the computation.

PiperOrigin-RevId: 219170087

er91 committed:
Add function to postprocess TF graph after encapsulating outside compilation.

PiperOrigin-RevId: 219170282

Commit:
Correctly insert converted_call even in non-recursive mode, because certain functions like built-ins need to be always processed.
This CL also removes the automatic wrapping of builtins to py_func, and calls them directly instead.

PiperOrigin-RevId: 219177870

Commit:
Avoid hardcoding the control variable to Tensor in break canonicalization.

PiperOrigin-RevId: 219185092

tensorflower-gardener committed:
Clarify the need for calling tensorflow::port::InitMain() before using the rest of TensorFlow.

PiperOrigin-RevId: 219185259

tanzhenyu committed:
Implement apply gradients for keras optimizer v2.

PiperOrigin-RevId: 219189655

tensorflower-gardener committed:
Delete out-of-date samples.

PiperOrigin-RevId: 219193262

tensorflower-gardener committed:
Speed up inference in presence of categorical features.

PiperOrigin-RevId: 219195789

Commit:
Fix log message in StreamExecutor, and add an explicit type in replay_computation.

Two minor changes:
 - The log message in deallocation in StreamExecutorAllocator prints out
   "(nil)" rather than the pointer because it happens after the pointer has
   been freed.

 - Add an explicit type in replay_computation; this was a review comment I missed.

PiperOrigin-RevId: 219195845

tensorflower-gardener committed:
Macro to print the names of the types when there's a mismatch.

PiperOrigin-RevId: 219196152

kayzhu committed:
[XLA] Fix the BUILD file license in XLA::Interpreter.

PiperOrigin-RevId: 219196383

Commit:
Allow ints to be converted to all floats.

PiperOrigin-RevId: 219196844

tensorflower-gardener committed:
Remove estimator_lib and import dependencies directly within estimator/__init__.py

PiperOrigin-RevId: 219198288

mkuperst committed:
[XLA] Delete OOM test.
It doesn't seem to serve a useful purpose, isn't actually enabled, and is a pain to maintain.

PiperOrigin-RevId: 219202422

tensorflower-gardener committed:
Merge pull request #23006 from girving:compatibility

PiperOrigin-RevId: 219199109

jdduke committed:
Add several trigonometric kernels to TFMobile

This includes:
 * cos
 * cosh
 * sin
 * sinh
 * tan

Note that tanh was already included, and TFLite already supports sin.

PiperOrigin-RevId: 219204972

tensorflower-gardener committed:
Automated rollback of commit 9a8e0c228dcf6a98d0f35b3737be9a07a43961ea

PiperOrigin-RevId: 219205308

annarev committed:
Automated rollback of commit fbd12e7e7b3c340ed82171610e40f063ef70ad26

PiperOrigin-RevId: 219208174

tensorflower-gardener committed:
Resolved some warnings in interpreter.h.

PiperOrigin-RevId: 219209314

ezhulenev committed:
Trim function library before starting Grappler passes

PiperOrigin-RevId: 219209842

tensorflower-gardener committed:
This CL automatically defuns the inner iteration of the eager Keras fit, evaluate, and predict loops if a model supports defuning. A graph network is marked as supporting defuns if all of its internal layers also support being in a defun. Subclass models and custom layers must be manually marked as supporting defuns in the constructor, by setting `self._can_use_graph_functions=True`

PiperOrigin-RevId: 219210710

tensorflower-gardener committed:
Fix rounding error with large tensors in tf.nn.zero_fraction.
The CPU kernel for reduce_mean does a sequential summation on eigen, which saturates.

PiperOrigin-RevId: 219210923

er91 committed:
Add function to extract outside compilation in XLA computations.

Unit test encapsulate_subgraphs_pass_test.cc now uses this function to extract outside compilation.

Two small changes w.r.t. original implementation:
1. Now we can infer output shapes in more cases (because we do shape inference earlier);
2. If there is control dependency between nodes in different outside compilations, do not add control dependency between their XlaHostCompute nodes.

PiperOrigin-RevId: 219211543

jdduke committed:
Whitelist trigonometric ops for flex conversion

PiperOrigin-RevId: 219214081

mrry committed:
[tf.data] Cache dataset structure properties in `ConcatenateDataset`.

Previously, these would be recomputed from the input on each call to
`ConcatenateDataset.output_shapes`, which would lead to pathological
behavior when building a long chain of `Dataset.concatenate()`
transformations.

PiperOrigin-RevId: 219217408

tensorflower-gardener committed:
Don't simplify nodes when output result from folding is too large on evaluation.

PiperOrigin-RevId: 219217834

tensorflower-gardener committed:
Remove DNNLinearCombinedEstimator from contrib

PiperOrigin-RevId: 219223773

froystig committed:
[XLA] Implement an optional XRT backend for the Python XLA client.

Also, consolidate XLA execute methods within the XLA client library backend.

PiperOrigin-RevId: 219224249

tensorflower-gardener committed:
Set _num_inputs property on _CuDNNRNN

PiperOrigin-RevId: 219225569

Commit:
Reject TF_XLA_FLAGS if any entries are not recognized.

Previously, we'd silently ignore unrecognized flags.

PiperOrigin-RevId: 219226600

yifeif committed:
Add Dockerfile for custom op.
This Dockerfile will be tagged as tensorflow/tensorflow:custom_op

PiperOrigin-RevId: 219227341

yifeif committed:
Clean license check so that it does not print irrelevant bazel ERROR msg.

PiperOrigin-RevId: 219227443

Commit:
Support Estimator.train and Estimator.evaluate to work with distribute coordinator and in-graph distribution strategy in standalone client mode.

It uses the whole cluster for training and evaluation and therefore it doesn't allow a "evaluator" job in cluster_spec.

PiperOrigin-RevId: 219227455

mrry committed:
[tf.data] Construct tensors in-place in the `out_tensors` vector whenever possible.

PiperOrigin-RevId: 219227733

tensorflower-gardener committed:
Typo in documentation.

PiperOrigin-RevId: 219227828

Commit:
Automated rollback of commit e0af1577de0f190fce023c792d60a9a2491ae8c3

PiperOrigin-RevId: 219228096

Commit:
Switch tensorflow::bit_cast -> absl::bit_cast.

PiperOrigin-RevId: 219228749

annarev committed:
Internal change.

PiperOrigin-RevId: 219229787

bixia1 committed:
[XLA:GPU] Remove the implicit reference of the last kernel in the thunk sequence
when ParallelLoopEmitter is used.

Currently, when IrEmitterUnnested handles an HLO for which ParallelLoopEmitter
is used to emit the kernel, we first create the thunk for the kernel, push the
thunk to the end of the thunk sequence and then call the handler for the same
HLO in IrEmitter. The handler in IrEmitter emits code to compute the base
pointers for parameters before constructing the elemental ir emitter, and then
passes the elemental ir emitter to IrEmitterUnnested::EmitTargetElementLoop,
which implicitly references the last kernel in the thunk and eventually invokes
ParallelLoopEmitter with the elemental ir emitter.

To remove such an implicit reference of the last kernel in the thunk sequence,
we postpone the creation of the thunk to routine
IrEmitterUnnested::EmitTargetElementLoop, and delay the emission of code that
computes the base pointers for parameters by wrapping the code emission inside
a generator and pass the generator to IrEmitterUnnested::EmitTargetElementLoop
through the FusedIrEmitter.
PiperOrigin-RevId: 219249592

shivaniag committed:
In accordance to changes in get_concrete_function.

PiperOrigin-RevId: 219256351

djvisentin committed:
Log error in _WrappedSession.close

PiperOrigin-RevId: 219257907

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2018-10-30

PiperOrigin-RevId: 219265416

Commit:
Add deps which cannot be added upstream because of circular dependencies.

PiperOrigin-RevId: 219270291

Commit:
Removed estimator/api BUILD file now that estimator has moved to new repo.

PiperOrigin-RevId: 219278964

Commit:
Cleaning up unused ESTIMATOR_API_INIT_FILES from api_gen.bzl

Estimator code has its own copy of this file. This list is no longer
needed in core tensorflow code.

PiperOrigin-RevId: 219280739

skye committed:
Make captured ResourceVariables work with while_v2 + gradients.

This involves a few changes:

* The gradients implementation doesn't error out if it gets a resource tensor y + grad_y (resource tensors previously only worked if grad_y was None). This is necessary since the while body function has matching resource tensor input/outputs, so we end up taking the gradient of the resource output, which has the default grad_y from the outer gradients call.

* We special-case setting the grad_y to None for TensorArrays only. This is currently done with a whitelist. If some TensorArray ops are missing from this list, we can add them later.

* We use the original input names in the body function signature. This isn't necessary but makes debugging easier (otherwise they all show up as "placeholder" and it's hard to tell which input is which).

* Adds new method func_graph.maybe_captured(tensor) to determine if a tensor represents a captured value.

PiperOrigin-RevId: 219298427

Commit:
[tf.data vectorization] Vectorizer for `Identity` op

PiperOrigin-RevId: 219300526

langmore committed:
linear_operator_util:  Special case some broadcasting for matmul and
triangular_solve.  This should speed these ops up in some cases.
PiperOrigin-RevId: 219302752

froystig committed:
[XLA] Support non-local XRT session targets in the Python XLA library.

PiperOrigin-RevId: 219306002

Commit:
[tf.data vectorization] Vectorizer for `Transpose`

PiperOrigin-RevId: 219307241

tensorflower-gardener committed:
Change RecvBufRespExtra.tensor_content to a repeated string and fill
it with many small strings instead of one large one, when using gRPC.

Typing tensor_content as a Cord instead of a single string leads to
roughly a 20% speedup in a 2-worker (8 v100 GPUs each) benchmark training of
resnet50 using collective all-reduce for gradient reduction and gRPC for
all inter-worker transport.  It is hypothesized that without the Cord
type gRPC is stalling incoming RecvBuf RPCs as it repeatedly reallocates
and copies the strings.  Using a Cord to receive the value leads to much
better flow control.

Unfortunately, proto3 does not yet support [ctype=CORD], so we can't
use that simple and effective optimization.  This CL changes
tensor_content to a sequence of strings and sets a max single-string
size of 4KB, the likely page size.  (This default can be changed via
ConfigProto.experimental.recv_buf_max_chunk.) It achieves roughly a
12% speedup on the benchmark test.  The speedups are highly dependent
on topology and network weather since the major effect is believed to
be on flow control.

PiperOrigin-RevId: 219322231

shashishekhar committed:
Fix build for android_png_internal.

PiperOrigin-RevId: 219323881

Commit:
[SE] Remove StreamExecutor alias for absl::InlinedVector.

PiperOrigin-RevId: 219324068

petewarden committed:
Reference implementation for audio preprocessing microcontroller example

PiperOrigin-RevId: 219326117

tensorflower-gardener committed:
Extend Identity optimizations to IdentityN.

PiperOrigin-RevId: 219327001

mkuperst committed:
[XLA] Avoid UB in absl::bitcast from Eigen::half.

PiperOrigin-RevId: 219331320

miaout17 committed:
Fix tflite_convert + Flex with bazel build.

PiperOrigin-RevId: 219331511

Commit:
[SE] Remove StreamExecutor alias for absl::string_view.

PiperOrigin-RevId: 219333558

tensorflower-gardener committed:
Automated rollback of commit 1034b710f925da02154f13ed75d7f4709c4d1d92

PiperOrigin-RevId: 219335805

tensorflower-gardener committed:
Utility to check if eager execution is enabled even if currently in a defun.

PiperOrigin-RevId: 219336026

tensorflower-gardener committed:
Fix inadvertently disabled test in templates_test

PiperOrigin-RevId: 219337792

shashishekhar committed:
Delete duplicate links.

PiperOrigin-RevId: 219338987

tensorflower-gardener committed:
Logged parameters in case of failure.

PiperOrigin-RevId: 219339372

annarev committed:
Add deprecation_version to api_def.proto and remove ops/endpoints in V2 that
have deprecation_version == 2.

PiperOrigin-RevId: 219343285

petewarden committed:
TensorFlow Lite for Micro documentation fixes

PiperOrigin-RevId: 219344543

timshen91 committed:
Bump opensource Abseil version.

PiperOrigin-RevId: 219344723

annarev committed:
Add a standard way to add a warning in tf_upgrade_v2.py script.

PiperOrigin-RevId: 219352016

annarev committed:
Add test_op.so back as data dependency to c_api_test. One of the functions it
tests is TF_LoadLibrary. I think we need to load a .so file to test TF_LoadLibrary functionality correctly. However, other tests can rely on static linking.

PiperOrigin-RevId: 219352122

tensorflower-gardener committed:
Missing test for type name function.

PiperOrigin-RevId: 219352921

Commit:
[tf.data] Fix flaky test from comparing floats

PiperOrigin-RevId: 219355162

timshen91 committed:
Refactor AlgorithmDesc and AlgorithmConfig.

The DNN library communicates to the caller with algorithm information in
3 ways:
(a) The caller asks DNN, for a list of available algorithms.
(b) When calling a convolution/RNN, the caller asks the DNN library for a suggestion
    of an algorithm. This also includes the case that the caller just wants one
    specific algorithm.
(c) The response of (b), sent from the DNN library back to the caller,
    including which algorithm is picked, and how much scratch space is used.

For (a), the data is (algo_id, tensor_ops_enabled).
For (b), the data is (optional (algo_id, tensor_ops_enabled), optional (algo_id, tensor_ops_enable) with no scratch).
For (c), the data is (selected (algo_id, tensor_ops_enabled), the actual scratch size).

Ideally we can have separate data structures for (a), (b) and (c), e.g.
AlgorithmDesc for (a), AlgorithmConfig for (b), and ProfileResult for
(c).

Currently, the scratch size is in AlgorithmDesc, and the optionality is
in AlgorithmDesc as well.

This patch moves scratch size to ProfileResult, and optionality to
AlgorithmConfig.

Historically, PR20708 first tried to add the scratch size to AlgorithmConfig,
but that's not the right place (as explained above). The scratch size ended up
in AlgorithmDesc, because I was concerned that changing ProfileResult was too much
work (I asked for a TODO for it, but I didn't find it in the current code).
Now hopefully all bits are moved to right places.

PiperOrigin-RevId: 219359118

tensorflower-gardener committed:
Add C API for use in the S4TF Compiler for fetching the number_attr
corresponding to an input argument list.

PiperOrigin-RevId: 219359725

er91 committed:
Propagate const input of "If"/"While" node into their function bodies.

Function call node and SymbolicGradient are already handled in a different way. When we compile function call node or SymbolicGradient, if some of their XLA inputs are XLA constant values, we will generate const HLO instead of parameter HLO (we effectively remove those const inputs from XLA computation's inputs). We could use the same mechanism for "If"/"While" nodes, but graph rewriting seems like a cleaner solution. Especially for "While" node: body function needs to have same input & output args; we need to perform graph rewriting if we remove Const input args.

Reason for not doing it in graph rewriting pass: these Const input nodes might only be available after constant folding. As a result, we can only do this after GraphOptimizer. Running GraphOptimizer in graph rewriting phase might break some graphs (e.g. GraphOptimizer removes Identity nodes in outside compilation which might be used as fetched tensor).

PiperOrigin-RevId: 219360183

yifeif committed:
Switch if_cuda back to if_cuda_is_configured as in PR #20277.

PiperOrigin-RevId: 219362371

Commit:
Make TFLite Python tests run in open source.

PiperOrigin-RevId: 219364657

hawkinsp committed:
[TF:XLA] Register device-to-device transfer streams in XlaDeviceContext.

PiperOrigin-RevId: 219365192

yifeif committed:
Move comment to top of command.

PiperOrigin-RevId: 219365298

fdxmw committed:
AsyncCheckpoints: Add missing 'self' arg to write_graph_fn.

PiperOrigin-RevId: 219365527

shivaniag committed:
Apply options for make_one_shot_iterator as well.

PiperOrigin-RevId: 219365879

Commit:
Don't auto-cluster stateful RNG ops

Auto-clustering stateful RNG ops is not safe because the seeding behavior of
TensorFlow stateful RNG ops is different that their seeding behavior in XLA so
auto-clustering these can introduce differences in behavior.

Once TF/XLA respects TensorFlow's seeding behavior, this CL can be reverted.

PiperOrigin-RevId: 219368223

tensorflower-gardener committed:
Support shrink_axis_mask in Grappler graph_properties.

PiperOrigin-RevId: 219372234

miaout17 committed:
Automated rollback of commit 3786aa194ce63010ac43413db49213629b3135db

PiperOrigin-RevId: 219372802

alextp committed:
Removes potential noisy error message from gradient tape.

PiperOrigin-RevId: 219374219

ezhulenev committed:
Give names to TensorId (std::pair) values.

PiperOrigin-RevId: 219376403

Commit:
Allow Slice ops with int64 indices on GPU

This will simplify cr/218907734 which is disabled on HEAD for now because it
creates slices with int64 indices which can't run on a GPU.

I also had to make Tile work for 64 bit indices to write a test case for Slice.
I did not add an independent test case for Tile since there isn't a
kernel_tests/tile_op_test.py but I'm happy to add one if you think it isn't an
overkill.

PiperOrigin-RevId: 219376726

tensorflower-gardener committed:
Internal change

PiperOrigin-RevId: 219378477

tensorflower-gardener committed:
Support TF <-> TFLite string conversion at the flex boundary (into the delegate)

PiperOrigin-RevId: 219381069

Commit:
Don't always convert and cast, as this can be slow.

PiperOrigin-RevId: 219383409

tensorflower-gardener committed:
Merge pull request #22606 from Intel-tensorflow:concat-primitive-descriptor-fix

PiperOrigin-RevId: 219385417

aaroey committed:
Remove logging of the GraphDef, it pollutes the log and make debugging harder.

PiperOrigin-RevId: 219386954

ezhulenev committed:
[Grappler] Do not copy unreachable functions into Grappler function item instantiation function library.

PiperOrigin-RevId: 219389437

jdduke committed:
Internal change

PiperOrigin-RevId: 219389755

tanzhenyu committed:
Enable hyper parameter tuning for optimizer V2.

PiperOrigin-RevId: 219390306

jsimsa committed:
[tf.data] Using consistent thread names.

PiperOrigin-RevId: 219390881

annarev committed:
Internal change.

PiperOrigin-RevId: 219390990

tensorflower-gardener committed:
Merge pull request #21466 from Intel-tensorflow:int8-part4

PiperOrigin-RevId: 219392938

lamberta committed:
Create landing page for XLA
Moving to top-level directory

PiperOrigin-RevId: 219395797

tensorflower-gardener committed:
Add support to quantize separable convolutions that use Resource Variables as weights.

PiperOrigin-RevId: 219396051

tensorflower-gardener committed:
Adds a function that returns quantization params for a given TFL_Tensor.

PiperOrigin-RevId: 219396856

annarev committed:
Internal change.

PiperOrigin-RevId: 219400038

tensorflower-gardener committed:
Mention support for multiple patterns in Dataset.list_files documentation. Add tests for it.

PiperOrigin-RevId: 219403096

shivaniag committed:
Copies attributes instead of moving for ParseExampleDataset.

In Eager mode, creating iterator again for the same dataset resulted in segmentation fault.
FIXES #23195

PiperOrigin-RevId: 219403151

er91 committed:
Add ExtractOutsideCompilationPass.

PiperOrigin-RevId: 219406439

jsimsa committed:
[tf.data] Refactoring of performance modeling implementation and adding performance modeling for all core and experimental tf.data kernels.

PiperOrigin-RevId: 219406929

saxenasaurabh committed:
Partial support for maximum_iterations in while_v2.
This does not support the XLA use-case yet since behavior of TensorListReserve needs to be changed.
Fix bug in while_v2_test.

PiperOrigin-RevId: 219411675

Commit:
Support tf.colocate_with when input is Tensor-like and executing_eagerly().

Graph.colocate_with applies the private function _op_to_colocate_with, which does tensor conversions if not already an Operation. This is not true for the TF op tf.colocate_with.

This specific edge case appears in the following link. Alternative solutions are welcome.

* Pull request: https://github.com/tensorflow/tensor2tensor/pull/1182/files#diff-c7da3e2b2808bb202eddd84075267eceR265
* Aforementioned parent method: https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/python/keras/engine/base_layer.py#L629

PiperOrigin-RevId: 219414649

mrry committed:
[Windows] Restore msvcp_dll_name attr in build_info.py when using the Bazel build.

PiperOrigin-RevId: 219416150

tensorflower-gardener committed:
[XLA] add .shape() to Python representation of device buffers in local client

PiperOrigin-RevId: 219422341

Commit:
Populate the internal autograph namespace in a more generic fashion.

PiperOrigin-RevId: 219423201

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2018-10-31

PiperOrigin-RevId: 219438285

tensorflower-gardener committed:
Allow function calls to be present inside pyct template replacements.

PiperOrigin-RevId: 219468133

tensorflower-gardener committed:
Improve unittest coverage of Reshape. (and add a bit more information to
a test_util message).

PiperOrigin-RevId: 219475399

tensorflower-gardener committed:
Clean up AutoGraph-generated pointers so that GC works correctly.

PiperOrigin-RevId: 219475499

aselle committed:
Move TensorFlow Lite Python into tensorflow.lite

Add tensorflow.contrib.lite into traversal of api generator.
+ Needed to change API generator to take list of deps and packages.
+ Exempt tensorflow.conrib.lite from METADATA dependency check (since
  we are moving)

PiperOrigin-RevId: 219483886

d0k committed:
[XLA:GPU] Use the scalar reduce algorithm for reductions to f32[1]

Any 1-sized tensor is physically identical to a scalar.

PiperOrigin-RevId: 219485837

ezhulenev committed:
[Grappler] Preserve constness of the GraphDef in GraphView.

1. Split GrapView into GraphView and MutableGraphView with separate {Input/Output}Port types with different node pointer constness.

2. Properly use GraphView and MutableGraphView in graph properties, and get rid of const_cast.

3. Remove const_cast in function optimizer.

4. Migrate GraphView to absl containers and hash

PiperOrigin-RevId: 219488040

jpienaar committed:
Only export colocate_with in TF V1.

PiperOrigin-RevId: 219489151

Commit:
TPUStrategy should not import from OneDeviceStrategy

PiperOrigin-RevId: 219491292

d0k committed:
[XLA:GPU] Output fuse bias adds into dot

TuplePointsToAnalysis already knows how to share the input to the add with the
output of the fusion. If for whatever reason the sharing is not possible, emit
a memcpy from input to output before calling gemm. The change is mostly
straight-forward, there's one complication with the online autotune of gemms.
We allocate a scratch buffer to avoid overwriting the bias during autotune.

This doesn't do anything fancy like trying to combine the scalar mul with the
add. We could do but it doesn't seem to be worth the additional complexity.

PiperOrigin-RevId: 219492278

ppham27 committed:
Fix typo in documentation for TPUEmbedding

Current documentation says optimization_parameters must not be `None`. The exception raised in the constructor says otherwise.

PiperOrigin-RevId: 219492950

tensorflower-gardener committed:
Don't rewrite `x / Sqrt(y)` -> `x * Rsqrt(y)` if `Sqrt(y)` is a fetch node.

PiperOrigin-RevId: 219496197

tensorflower-gardener committed:
Optimize EdgeSet data structure:

1. Extend the inlined part to use an entire cacheline.
2. Switch from std::set to gtl::FlatSet (absl::flat_hash_set breaks the android build), which uses less memory, causes less memory fragmentation and is faster.

This speeds up edge iteration on large nodes by up to 80%.
This speeds up GraphDef->Graph conversion by up to 44%.
Saves about 20% of memory for nodes that spill over the inlined set.

PiperOrigin-RevId: 219508843

annarev committed:
Include stdlib.h in contrib/lite/string_util.cc.

PiperOrigin-RevId: 219509233

tensorflower-gardener committed:
Changing documentation regarding batch_normalization update operations.

PiperOrigin-RevId: 219510417

dubey committed:
Override GPU order at each worker via ConfigProto for CollectiveOps.

While collectives includes logic for automatically deducing a good ring order
among devices on each worker, after this change the user can override the ring
order via the session configuration.  If specified, this defines the order of
GPU devices along which data is exchanged at each worker.  This assumes that
the GPU topology is identical at each worker, and hence the same ring order is
good for all workers.

The ring order in ConfigProto is used for all collective ops in the session.
Multiple session.run calls will reuse the same ring order.

PiperOrigin-RevId: 219510944

Commit:
[XLA] BF16 passes: avoid changing side-effecting instructions.

PiperOrigin-RevId: 219514687

shivaniag committed:
Optimize on calling get_concrete_function() multiple times.

PiperOrigin-RevId: 219518529

shivaniag committed:
Automated rollback of commit 0866dadf1d20b6edd3bfb1a0b01409fc1a634c9a

PiperOrigin-RevId: 219518544

jsimsa committed:
[tf.data] Fix an internal test failure.

PiperOrigin-RevId: 219520598

annarev committed:
Removing deprecated ApiDef endpoints in V2. Removal is accomplished by
by setting deprecation_version.

PiperOrigin-RevId: 219524292

Commit:
[tf.data vectorization] vectorization bug fix to prevent infinite recursion

PiperOrigin-RevId: 219524470

mrry committed:
[tf.data] Optimize the performance of `tf.data.Iterator` in eager mode.

The change to introduce `Structure` into iterators regressed the per-element performance because it added redundant checks each time an element was produced. With this change, the performance is slightly better than it was before the move to `Structure`.

PiperOrigin-RevId: 219532162

ispirmustafa committed:
Automated rollback of commit 6695dab0104e311c9221cd47bb8685d840709da2

PiperOrigin-RevId: 219532483

tensorflower-gardener committed:
Fix timeout in edgeset_test.

PiperOrigin-RevId: 219533560

angerson committed:
Migrate TensorFlow Lite out of tensorflow/contrib

This change moves //tensorflow/contrib/lite to //tensorflow/lite in preparation
for TensorFlow 2.0's deprecation of contrib/. If you refer to TF Lite build
targets or headers, you will need to update them manually. If you use TF Lite
from the TensorFlow python package, "tf.contrib.lite" now points to "tf.lite".
Please update your imports as soon as possible.

For more details, see https://groups.google.com/a/tensorflow.org/forum/#!topic/tflite/iIIXOTOFvwQ

@angersson and @aselle are conducting this migration. Please contact them if
you have any further questions.

PiperOrigin-RevId: 219536476

tensorflower-gardener committed:
Automated rollback of commit f6b8d21537595bb947e9dfd25c082a56722f5483

PiperOrigin-RevId: 219536671

shivaniag committed:
[data-stats] Collects `active_parallel_calls` and `num_parallel_calls` as scalar, and `parallel_calls_utilization` as histogram for ParallelMapDataset, MapAndBatchDataset, ParallelInterleaveDatasetV2,and ParseExampleDataset.

PiperOrigin-RevId: 219539778

ezhulenev committed:
Add FunctionLibraryRuntime overlay for use in executors created for function instantiations with overlay libs.

+ enable Grappler function optimizer in tf-eager

PiperOrigin-RevId: 219545431

nfelt committed:
Internal change.

PiperOrigin-RevId: 219546447

tensorflower-gardener committed:
Merge pull request #23304 from feihugis:Issue-Minor_Update_Log_Format

PiperOrigin-RevId: 219552367

tensorflower-gardener committed:
Fix ASAN error.

PiperOrigin-RevId: 219555843

bixia1 committed:
[XLA] Make PropagateBufferConstraint and PropagateOperandConstraint in the
common layout assignment pass propagate layouts for non-layout changing
instructions.

This also fixes the problem discovered in b/117531509.

Turn on the check for non-layout-changing instructions should not change
layouts in the GPU backend.

Add a test case.

PiperOrigin-RevId: 219558979

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2018-11-01

PiperOrigin-RevId: 219603292

Commit:
[TF:XLA] Bump open source abseil revision to cc8dcd307b76a575d2e3e0958a4fe4c7193c2f68

PiperOrigin-RevId: 219606447

hawkinsp committed:
[TF:XLA] Change implementation of ReverseSequence to use an XLA gather rather than an explicit loop.

PiperOrigin-RevId: 219631839

hawkinsp committed:
[TF:XLA] Unconditionally use xla::TransferManager to transfer buffers on XLA_* devices, and remove the non-TransferManager path.

This change only affects pure XLA devices like XLA_CPU and XLA_GPU, and does not affect XLA-compiled code running on the regular CPU and GPU devices (i.e., most users). There's little downside to doing this now that we have the LiteralSlice/BorrowingLiteral XLA APIs which don't require unnecessary copies.

Also clarifies the semantics of when host-side data structures can be freed in xla::TransferManager::TransferLiteralToDevice, and simplifies the corresponding code in XlaDeviceContext.

PiperOrigin-RevId: 219644033

allenlavoie committed:
Tweak to colocate_with when executing eagerly to avoid variable reads

Employs duck typing to fetch a "device" attribute rather than eagerly converting to a Tensor.

PiperOrigin-RevId: 219646805

Artem-B committed:
Make NVPTX logging somewhat more structured.

-v=2 enables 'logistics' logs -- tool/library search, PTX/CUBIN size info.
-v=3 enables detailed printout of IR/PTX/thunk schedules.
PiperOrigin-RevId: 219649093

tensorflower-gardener committed:
Add support for strings in Reshape.

PiperOrigin-RevId: 219650781

tensorflower-gardener committed:
Enable lite/testing tests

PiperOrigin-RevId: 219651202

lamberta committed:
Fix XLA _book.yaml.
Fix lite and update subsite config to new location.
Publish other parts of the site.

PiperOrigin-RevId: 219652989

petewarden committed:
Contrib migration fixes for microcontroller build

PiperOrigin-RevId: 219653425

allenlavoie committed:
Add variable support to 2.x SavedModel export

This is still focused on exporting to Session-based import APIs. Should be ready to add to the 2.x preview after this and possibly a few follow up fixes (for the reference cycle issue in the skipped unit test).

Maps resources which are owned by the eager context to new resources in the Graph being exported.

PiperOrigin-RevId: 219655229

MarkDaoust committed:
Add version note to lite/python_api page.

PiperOrigin-RevId: 219655764

superbobry committed:
Warn on add_* calls on a closed tf.summary.FileWriter.

Once the writer is closed, all the add_* methods become noops. This
behaviour could result in silent bugs, e.g.

    writer = tf.summary.FileWriter(...)
    for _ in range(train_steps):
      summary = sess.run(...)
      writer.add_summary(summary)

      # ... more code here followed by *misindented*
      writer.close()

PiperOrigin-RevId: 219655902

tensorflower-gardener committed:
Restore input_test.
Just disable the part of the tests that were flaky instead of the whole test.

PiperOrigin-RevId: 219660130

Commit:
Fix typo in comment

PiperOrigin-RevId: 219660859

aaroey committed:
Verify the node before putting it into a TRT subgraph, so if the TRT converter
doesn't support it, it can be excluded earlier. This prevents a single
incompatible node from causing the conversion of a large subgraph to fail.

Currently this is only added for Transpose, Reshape and Const.

Implementation details:

- A new class TrtCandidateSelector is added for the validation. Previously it
  used IsTensorRTCandidate() to pick TRT candidate nodes which only checks the
  op type; the new TrtCandidateSelector::IsTensorRTCandidate() checks the op
  type as well as input properties of the node, e.g. const or not, the values of
  the const, etc.
- TrtCandidateSelector uses TrtNodeValidator added earlier to validate the
  nodes. A new method TrtNodeValidator::GetTensorOrWeights() is added to
  convert an input of a node to a TRT_TensorOrWeights, and after gathering all
  input TRT_TensorOrWeights, TrtNodeValidator::ValidateNode() is called to
  determine whether the node is supported by TRT.
- The original InputEdgeValidator class is now removed. It only validates the
  input nodes of the subgraph, and now its logic is merged into
  ValidateTensorProperties which will be used by TrtNodeValidator to validate
  all nodes before putting them in subgraph.

PiperOrigin-RevId: 219662352

meheffernan committed:
Verify parameter count of called computation in HloVerifier.

PiperOrigin-RevId: 219662775

allenlavoie committed:
Fix some reference cycles in TF 2.x SavedModel export

There are still some in Python 2.7

PiperOrigin-RevId: 219667378

hawkinsp committed:
[TF:XLA] Use the correct function library in ondemand compilation op.

Enable tests that were broken because they couldn't find functions.

PiperOrigin-RevId: 219667993

tensorflower-gardener committed:
Merge pull request #22734 from mslacken:master

PiperOrigin-RevId: 219667359

tensorflower-gardener committed:
Fix path to flatc in visualize.py

PiperOrigin-RevId: 219669159

tensorflower-gardener committed:
Add arbitrary axis support to Gather

PiperOrigin-RevId: 219673298

Commit:
[XLA] BF16 passes: use HasSideEffectNoRecurse

HasSideEffect is too conservative, since we visit all individual HLOs.

PiperOrigin-RevId: 219678454

hawkinsp committed:
[TF:XLA] Relax tolerance of FTRL test.

PiperOrigin-RevId: 219684749

gmagogsfm committed:
Disable estimator_training_test in {a,m,t}san

PiperOrigin-RevId: 219686752

tensorflower-gardener committed:
Removing nanoproto support.

PiperOrigin-RevId: 219687136

kayzhu committed:
[TF2XLA] Fix bilinear_resize to correctly specify batch dimension size when
slicing from input.

PiperOrigin-RevId: 219688228

k-w-w committed:
Remove SavedModel functions from the 2.0 API.

PiperOrigin-RevId: 219688691

tensorflower-gardener committed:
Merge pull request #23175 from Intel-tensorflow:bhavanis/matmul-bug-fix

PiperOrigin-RevId: 219688787

fchollet committed:
Use correct TensorShape class in V2.

PiperOrigin-RevId: 219692084

Commit:
Remove CollectiveAllReduceStrategy as eval strategy in estimator_training_test.

PiperOrigin-RevId: 219692628

denisvnukov committed:
[XLA] Extended execution profiler to allow extra metrics.

Also added entry computation name to HloProfilePrinterData,
so that the printer can identify entry computation.

PiperOrigin-RevId: 219692668

fchollet committed:
Make sure None can be deserialized as None by the Keras serialization utils.

PiperOrigin-RevId: 219693744

jtkeeling committed:
Improve error message when no kernel found

This error occurs when an Op node does not have a registered kernel that supports the requested attributes. This improves the error message to:

* Add rich formatting that allows the Python layer to fill in where the node was created
* Specify the (implicitly) requested attributes, so the user can more easily see why none of the the listed kernels are suitable

PiperOrigin-RevId: 219694648

tensorflower-gardener committed:
When building NCCL from source, do not rely on CUDA_TOOLKIT_PATH being defined to find bin2c.

PiperOrigin-RevId: 219698198

nutsiepully committed:
Copying tensorflow contrib quantize to new model optimization repository.

PiperOrigin-RevId: 219700359

Commit:
[TF:XLA] Bump open source llvm revision to r345798

PiperOrigin-RevId: 219704389

mrry committed:
[tf.data] Fix a memory leak of `TrackingAllocator` objects when stats aggregation or modeling is enabled.

Setting a stats collector on an executor automatically enables
allocation tracking, which creates a `TrackingAllocator` in each
`OpKernelContext` for an op that allocates tensors.  The lightweight
`StepStatsCollectorInterface` implementation in tf.data was missing a
heretofore undocumented deallocation call for these allocator objects,
and so they were leaking.

PiperOrigin-RevId: 219709003

tensorflower-gardener committed:
Register Reverse & ReverseV2 for T=bool on GPU.

PiperOrigin-RevId: 219711803

Commit:
Fix how IncreaseDynamismForAutoJitPass was emitting the _XlaCompileTimeConstantInputs

_XlaCompileTimeConstantInputs is a list of input names, not input indices.  I
will add an integration test when I re-enable IncreaseDynamismForAutoJitPass to
make sure this is tested end to end.

PiperOrigin-RevId: 219715448

tensorflower-gardener committed:
Don't use the numerically unstable MeanReducer class in Eigen.

PiperOrigin-RevId: 219715636

miaout17 committed:
Rename flex_ops -> select_tf_ops in Toco flags

PiperOrigin-RevId: 219719783

tensorflower-gardener committed:
Adds checks for post pruning and sparse features when using oblivious trees.

PiperOrigin-RevId: 219720126

tensorflower-gardener committed:
Clean up, less nesting of nested if statements.

PiperOrigin-RevId: 219721219

Commit:
Dump GraphDef before increase_dynamism_for_auto_jit_pass

PiperOrigin-RevId: 219721816

gmagogsfm committed:
Internal changes.

PiperOrigin-RevId: 219723198

rxsang committed:
Fix a bug in tpu.py and xla.py that while creating an identity node for control input edges under rewrite context, the parent control flow context is lost.

PiperOrigin-RevId: 219724472

gmagogsfm committed:
Internal changes

PiperOrigin-RevId: 219726466

tanzhenyu committed:
Implement get_weights and set_weights for V2 optimizer.

PiperOrigin-RevId: 219731431

saxenasaurabh committed:
Support serializing tensor list with invalid tensors. This can happen if a TL is created using TensorListReserve.

PiperOrigin-RevId: 219736904

miaout17 committed:
Trying to fix Flex Converter.

PiperOrigin-RevId: 219737933

Commit:
Allow slicing from DT_INT64 tensors

IncreaseDynamismForAutoJitPass creates these and supporting them seems like the
right thing to do.

PiperOrigin-RevId: 219738687

rachellj218 committed:
Replace tf.keras.callbacks.TensorBoard callback with a new callback TensorBoardWithValidation inherited from the former callback, which makes evaluation at the end of specified epochs and export the results to tensorboard.

PiperOrigin-RevId: 219749605

tensorflower-gardener committed:
Export 'configure' and 'configure.py' to allow inverse dependent repos to do 'bazel run @org_tensorflow//:configure -- --workspace $(pwd)'.

PiperOrigin-RevId: 219764425

tensorflower-gardener committed:
compat: Update forward compatibility horizon to 2018-11-02

PiperOrigin-RevId: 219767338

Commit:
Internal Change.

PiperOrigin-RevId: 219796522

tensorflower-gardener committed:
toco: Pass over binary fusion when model is already quant.
PiperOrigin-RevId: 219796818

hawkinsp committed:
[TF:XLA] Implement LeakyRelu and LeakyReluGrad.

PiperOrigin-RevId: 219802754

tensorflower-gardener committed:
Improve error messages when types are not supported.

PiperOrigin-RevId: 219804668

tensorflower-gardener committed:
Automated rollback of commit 26376a4430cb109af2c6b41ffff8b8642799dffe

PiperOrigin-RevId: 219805304

tensorflower-gardener committed:
Update ops-related pbtxt files.

PiperOrigin-RevId: 219809618

georgesterpu committed:
demo
